modules/gem5/_posts/2000-05-01-assignment4.md

```diff
@@ -0,0 +1,331 @@
+I.e. matrices are indexed like `A[row number][column number]`.
+The starter code first iterates over the rows of `A` (*i*), then iterates over the columns of `B` (*j*), and then iterates over the elements in the selected row and column (*k*).
+
+```cpp
+void multiply(double **A, double **B, double **C, int size)
+{
+    for (int i = 0; i < size; i++) {
+        for (int j = 0; j < size; j++) {
+            for (int k = 0; k < size; k++) {
+                C[i][j] += A[i][k] * B[k][j];
+            }
+        }
+    }
+}
+```
+
+The animation below shows how the order of operations during the multiplication operation.
+
+![matrix multiplication]({{ '/img/mm_ijk.gif' | relative_url }})
+
+You can import this workload into your configuration script from `workloads/matmul_workload.py` as `IJKMatMulWorkload`.
+
+I have provided a ready made binary for this implementation.
+However, if you need to built this binary at any point, you can run the command below in `workloads/matmul`.
+
+```shell
+make mm-ijk-gem5
+```
+
+**CAUTION**: the above command generates a binary that could only be run with gem5.
+Running this binaries on your host will result in multiple errors.
+
+### A stationary matrix multiplication (ikj)
+
+As you can see in the animation, the previous implementation of the matrix multiplication does not expose much locality in its accesses to matrices `A` and `B`.
+Rather, it exposes locality when accessing matrix `C`.
+We can reorder our nested loops and still get the correct answer.
+In fact all permutations of our three nested loops will result in the same `C` matrix.
+Moreover, when it comes to the complexity of the algorithm itself, reordering the for loops does not change the complexity.
+However, reordering the for loops, changes the access pattern to the memory.
+In turn, it will increase/decrease our cache hit rate.
+For this step, you can see a slightly different implementation of the matrix multiplication program below.
+You can also find it in `workloads/matmul/ikj_multiply.h`
+
+```cpp
+void multiply(double **A, double **B, double **C, int size)
+{
+    for (int i = 0; i < size; i++) {
+        for (int k = 0; k < size; k++) {
+            for (int j = 0; j < size; j++) {
+                C[i][j] += A[i][k] * B[k][j];
+            }
+        }
+    }
+}
+```
+
+As you can see, two for loops have been interchanged (*j* and *k*) from our previous implementation.
+The animation below shows how the order of operations during the multiplication operation.
+
+![matrix multiplication]({{ '/img/mm_ikj.gif' | relative_url }})
+
+You can import this workload into your configuration script from `workloads/matmul_workload.py` as `IKJMatMulWorkload`.
+
+I have provided a ready made binary for this implementation.
+However, if you need to built this binary at any point, you can run the command below in `workloads/matmul`.
+
+```shell
+make mm-ikj-gem5
```

Comment:
General feedback (no need to take any action): Try to be consistent in how you name things. For instance the workload is "IKJMatMul" but here you call the binary "mm-ikj" (a slightly different name). This can lead do confusion and extra congnitive load for your readers/reviewers/students.